python/dgl/graphbolt/utils/sample_utils.py

```diff
@@ -52,44 +53,55 @@ def unique_and_compact_node_pairs(
     {('n1', 'e1', 'n2'): (tensor([0, 1, 1]), tensor([0, 1, 0])),
     ('n2', 'e2', 'n1'): (tensor([0, 1, 0]), tensor([0, 1, 1]))}
     """
-    is_homogeneous = not isinstance(node_pairs, Dict)
+    is_homogeneous = not isinstance(node_pairs, dict)
+
     if is_homogeneous:
         node_pairs = {("_N", "_E", "_N"): node_pairs}
-    nodes_dict = defaultdict(list)
-    # Collect nodes for each node type.
-    for etype, node_pair in node_pairs.items():
-        u_type, _, v_type = etype
-        u, v = node_pair
-        nodes_dict[u_type].append(u)
-        nodes_dict[v_type].append(v)
 
-    unique_nodes_dict = {}
-    inverse_indices_dict = {}
-    for ntype, nodes in nodes_dict.items():
-        collected_nodes = torch.cat(nodes)
-        # Compact and find unique nodes.
-        unique_nodes, inverse_indices = torch.unique(
-            collected_nodes,
-            return_inverse=True,
-        )
-        unique_nodes_dict[ntype] = unique_nodes
-        inverse_indices_dict[ntype] = inverse_indices
+    # Collect all source and destination nodes.
+    src_nodes = defaultdict(list)
+    dst_nodes = defaultdict(list)
+    for etype, (src_node, dst_node) in node_pairs.items():
+        src_nodes[etype[0]].append(src_node)
+        dst_nodes[etype[2]].append(dst_node)
+    src_nodes = {ntype: torch.cat(nodes) for ntype, nodes in src_nodes.items()}
+    dst_nodes = {ntype: torch.cat(nodes) for ntype, nodes in dst_nodes.items()}
+    # Compute unique destination nodes if not provided.
```

Comment:
nit: Is it slightly better to do this in C++ code?

Reply:
This preprocess is all about tensor op, I want to keep C++ code small and preformance concentrated.